pytorch_lightning/accelerators/mps.py

```diff
@@ -0,0 +1,98 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Dict, List, Optional, Union
+
+import torch
+
+from pytorch_lightning.accelerators.accelerator import Accelerator
+from pytorch_lightning.utilities import device_parser
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.imports import _PSUTIL_AVAILABLE, _TORCH_GREATER_EQUAL_1_12
+from pytorch_lightning.utilities.types import _DEVICE
+
+if _TORCH_GREATER_EQUAL_1_12:
+    _MPS_AVAILABLE = torch.backends.mps.is_available()
+else:
+    _MPS_AVAILABLE = False
+
+
+class MPSAccelerator(Accelerator):
+    """Accelerator for Metal Apple M1 GPU devices."""
+
+    def setup_environment(self, root_device: torch.device) -> None:
+        """
+        Raises:
+            MisconfigurationException:
+                If the selected device is not MPS.
+        """
+        super().setup_environment(root_device)
+        if root_device.type != "mps":
+            raise MisconfigurationException(f"Device should be MPS, got {root_device} instead.")
+
+    def get_device_stats(self, device: _DEVICE) -> Dict[str, Any]:
+        """Get M1 (cpu + gpu) stats from ``psutil`` package."""
+        return get_device_stats()
+
+    @staticmethod
+    def parse_devices(devices: Union[int, str, List[int]]) -> Optional[List[int]]:
+        """Accelerator device parsing logic."""
+        parsed_devices = device_parser.parse_gpu_ids(devices, include_mps=True)
+        return parsed_devices
+
+    @staticmethod
+    def get_parallel_devices(devices: Union[int, str, List[int]]) -> List[torch.device]:
```

Comment:
This method does not return an optional but `patse_devices` does. I think they should both do it or not do it for consistency

Reply:
I just sticked to the interface the other accelerators had. If I modify the typing here, wouldn't that violate liskov substitution principle?

Reply:
I don't see why. The base indicates `Any` anyways.

Feel free to resolve it as my comment is more about why does https://github.com/PyTorchLightning/pytorch-lightning/blob/dd475183227644a8d22dca3deb18c99fb0a9b2c4/pytorch_lightning/utilities/device_parser.py#L88-L90 return `None` instead of `[]`, which would remove the assert you had to add in the MPSAccelerator

Reply:
I agree that this case could be handled in the accelerator connector as well. But I think it is rather device specific, so I also understand why it is not there.
Maybe this should be resolved and me moved outside to be checked beforehand separately :)